compiler/src/dotty/tools/dotc/typer/RefChecks.scala

```diff
@@ -782,34 +782,53 @@ object RefChecks {
         report.error(problem(), clazz.srcPos)
       }
 
-      // check that basetype and subtype agree on types of trait parameters
+      // Checks base types of the current clazz against each other
       //
-      // I.e. trait and class parameters not only need to conform to the expected
-      // type of the corresponding base-trait, but also to the type as seen by the
-      // inheriting subtype.
-      def checkTraitParametersOK() = for {
-        parent <- clazz.info.parents
-        parentSym = parent.classSymbol
-        if parentSym.isClass
-        cls = parentSym.asClass
-        if cls.paramAccessors.nonEmpty
-        param <- cls.paramAccessors
-      } {
-        val tpeFromParent = parent.memberInfo(param)
-        val tpeFromClazz = clazz.thisType.memberInfo(param)
-        if (!(tpeFromParent <:< tpeFromClazz)) {
-          val msg =
-            em"""illegal parameter: The types of $param do not match.
-                |
-                |  $param in $cls has type: $tpeFromParent
-                |  but $clazz expects $param to have type: $tpeFromClazz"""
-
-          report.error(msg, clazz.srcPos)
+      // In particular, it checks that there are no two base classes with
+      // different type instantiations.
+      //
+      // ported from Scala 2:
+      //   https://github.com/scala/scala/blob/9bb659e62a9239c01aec14c171f8598bb1a576fe/src/compiler/scala/tools/nsc/typechecker/RefChecks.scala#L834-L883
+      def validateBaseTypes(): Unit = {
+        val tpe = clazz.thisType // in Scala 2 this was clazz.tpe
+        val seenParents = mutable.HashSet[Type]()
+        val baseClasses: List[ClassSymbol] = clazz.info.baseClasses
+
+        // tracks types that we have seen for a particular base class in baseClasses
+        val seenTypes = mutable.Map.empty[Symbol, List[Type]]
```

Comment:
The original Scala 2 code used an Array since baseClasses where stored in a `Seq` and allowed indexing into the array. Do you think it is a performance problem to use a `Map` here instead?

Reply:
Performance of `register` was mentioned here: https://github.com/scala/scala/commit/42fb66a2cbcad4fff0634dc935f135494f04be93

Reply:
We have a specialized MutableSymbolMap that you probably want to use here.

Reply:
An alternative to using a map here would be to use a set which should be good enough to detect one error and stop.

Reply:
Looks like you tried something like this but ended up reverting it: https://github.com/lampepfl/dotty/compare/2201e3da859ced8edb9da7a71a931cd2bc5cc395..fe73b0b5371d5b4181c1a98be2dc881981c0b526, was there an issue with it?

Reply:
Yes, it actually was not semantics preserving and caused some regressions -- So I just reverted it.